crates/anvil-polkadot/src/api_server/trace_helpers.rs

```diff
@@ -0,0 +1,145 @@
+/// This is used to build the output for Parity client's RPC method `trace_transaction`.
+pub fn parity_transaction_trace_builder(
+    trace: Trace,
+    tx_info: Option<TransactionInfo>,
+) -> Result<Vec<LocalizedTransactionTrace>, Error> {
+    let call_trace = match trace {
+        Trace::Call(call_trace) => call_trace,
+        Trace::Prestate(_) => {
+            return Err(Error::InternalError("Trace is not a call trace".to_string()));
+        }
+    };
+    let mut parity_tx_traces = Vec::new();
+    let mut next_traces = vec![(vec![], call_trace)];
+    while let Some((trace_address, trace)) = next_traces.pop() {
+        let transaction_trace =
+            parity_transaction_trace_from_call_trace(trace.clone(), trace_address.clone())?;
+        let localized_trace = LocalizedTransactionTrace {
+            trace: transaction_trace,
+            block_hash: tx_info
+                .as_ref()
+                .map(|tx_info| B256::from_slice(tx_info.block_hash.as_ref())),
+            block_number: tx_info
+                .as_ref()
+                .map(|tx_info| tx_info.block_number.try_into().unwrap_or_default()),
+            transaction_hash: tx_info
+                .as_ref()
+                .map(|tx_info| B256::from_slice(tx_info.hash.as_ref())),
+            transaction_position: tx_info
+                .as_ref()
+                .map(|tx_info| tx_info.transaction_index.try_into().unwrap_or_default()),
+        };
+        parity_tx_traces.push(localized_trace);
+        for (call_index, call) in trace.calls.iter().enumerate() {
+            let mut new_trace_address = trace_address.clone();
+            new_trace_address.push(call_index);
+            next_traces.push((new_trace_address, call.clone()));
+        }
+    }
+    Ok(parity_tx_traces)
+}
+
+/// Builds a Parity TransactionTrace from a CallTrace from pallet revive and a trace address,
+/// which is built with the indices of the path from the transaction root callto the specific call
+/// in the call tree.
+fn parity_transaction_trace_from_call_trace(
+    trace: CallTrace,
+    trace_address: Vec<usize>,
+) -> Result<TransactionTrace, Error> {
+    match trace.call_type {
+        ReviveCallType::Call | ReviveCallType::StaticCall | ReviveCallType::DelegateCall => {
+            Ok(TransactionTrace {
+                action: Action::Call(CallAction {
+                    from: Address::from_slice(trace.from.as_ref()),
+                    call_type: CallType::Call,
+                    gas: trace.gas.try_into().map_err(|_| EthRpcError::ConversionError)?,
+                    input: trace.input.0.into(),
+                    to: Address::from_slice(trace.to.as_ref()),
+                    value: AlloyU256::from(trace.value.unwrap_or_default()).inner(),
+                }),
+                error: trace.error,
+                result: Some(TraceOutput::Call(CallOutput {
+                    gas_used: trace.gas.try_into().map_err(|_| EthRpcError::ConversionError)?,
```

Comment:
Should we use trace.gas_used here?
```suggestion
                    gas_used: trace.gas_used.try_into().map_err(|_| EthRpcError::ConversionError)?,
```